README.md
# Cloudflare Helper
A very primitive MyBB plugin for exploiting HTTP headers set by Cloudflare.

## Features
This plugin detects HTTP header fields that are [added by Cloudflare](https://support.cloudflare.com/hc/en-us/articles/200170986-How-does-Cloudflare-handle-HTTP-Request-headers-) when a request is made through the Cloudflare network. Relevant headers are as follows:
- `CF-IPCountry`
- `CF-Connecting-IP`
- `X-Forwarded-For`
- `X-Forwarded-Proto`
- `CF-RAY`
- `CF-Visitor`
- `CDN-Loop`

Currently the plugin filters requests made to HTTP server by the `$_SERVER["REMOTE_ADDR"]` variable. If the value of `$_SERVER["REMOTE_ADDR"]` is in [Cloudflare IP Ranges](https://www.cloudflare.com/ips/), the plugin will process relevant [hooks](#hooks).

Object `$mybb_plugin_cloudflare_helper` holds relevant properties of the plugin. Developers could call its public functions to get above and other useful variables.

_**Note:**_ If your web server is behind a front-end web server, e.g. a reversed proxy server, the value of `$_SERVER["REMOTE_ADDR"]` the plugin gets is the front-end web server's IP and it'll recognize the requests are made from outside the Cloudflare's network. However in this circumstance, `X-Forwarded-For` could be used for filtering requests, maybe a feature version would support this.

## Requirements

- MyBB 1.8.x
- PHP >= 5.6
- The plugin doesn't need Cloudflare's proxy to be enabled for your forum. But without Cloudflare's proxy, it's pretty useless.

## Version & Upgrade notice

The initial version is `0`. Following versions will be updated when features are added and bugs are fixed. **It's only for test at this moment.**

For upgrade notice, please read [Upgrade/Reinstall Notice](UPGRADE.md).

## Installation

1. Upload all files/folders under the `Upload` directory to your MyBB root folder. Please maintain the folder structure within the `Upload` directory.
2. Activate plugins "Cloudflare Helper" and "Cloudflare Enhancer" at your MyBB's Admin Control Panel.

## Hooks

Following hooks are attached to MyBB's `global_start` hook:
- `cloudflare_helper_global_start`\
  `cloudflare_helper_global_start_end`

- `cloudflare_helper_global_start_NOTFROMCF`\
  `cloudflare_helper_global_start_end_NOTFROMCF`\
  The request is made from outside the Cloudflare network.
  
- `cloudflare_helper_global_start_NOCOUNTRY`\
  `cloudflare_helper_global_start_end_NOCOUNTRY`\
  Cloudflare can't detect the country where the request is made from. See also [Configuring Cloudflare IP Geolocation](https://support.cloudflare.com/hc/en-us/articles/200168236-What-does-Cloudflare-IP-Geolocation-do-).
  
- `cloudflare_helper_global_start_FROMTOR`\
  `cloudflare_helper_global_start_end_FROMTOR`\
  The request is made from Tor network through the Cloudflare network.
  
- `cloudflare_helper_global_start_{CC_CODE}`\
  `cloudflare_helper_global_start_end_{CC_CODE}`\
  The request comes from a country with country code [`{CC_CODE}`](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2). See [Configuring Cloudflare IP Geolocation](https://support.cloudflare.com/hc/en-us/articles/200168236-What-does-Cloudflare-IP-Geolocation-do-).

## TODO

- ~Switch to use `$_SERVER["REMOTE_ADDR"]` to check if a access request to MyBB comes from Cloudflare, since HTTP headers can be spoofed. See [Cloudflare IP Ranges](https://www.cloudflare.com/ips/).~
- Add hooks that would be executed when requests are not from a specific country.
- Add relevant PHPDocs.
- (still planning) Add settings for "Cloudflare Enhancer" plugin in "Board Settings".
- Add country code to country name mapping. (Member function `get_country()` is not implemented.)
- Auto-update for Cloudflare's IP ranges.

## About "Cloudflare Enhancer"

The "Cloudflare Enhancer" plugin is a demo to show how this Cloudflare Helper works by displaying some information below MyBB's debug stuff (normally at the page bottom).

_**Note:**_ You need to enable **"Advanced Stats / Debug information"** (in "Board Settings >> Server and Optimization Options" setting group) for MyBB to show MyBB's debug stuff.

## License
See the [LICENSE](https://github.com/yuliu/mybb-plugin-cloudflare-helper/blob/master/LICENSE) file.

## Huh?
...

CHANGELOG.md
# Changelog

## (on-going) v0.1

- Code refactor

- Switch to use `$_SERVER["REMOTE_ADDR"]` to check if a access request to MyBB comes from Cloudflare, since HTTP headers can be spoofed. See [Cloudflare IP Ranges](https://www.cloudflare.com/ips/). 

## v0 Initial release

Initial release.

Upload/inc/plugins/cloudflare_enhancer.php
<?php
/**
 * MyBB 1.8 plugin: Clouldflare Enhancer
 * Website: https://github.com/yuliu/mybb-plugin-cloudflare-helper
 * License: https://github.com/yuliu/mybb-plugin-cloudflare-helper/blob/master/LICENSE
 * Copyright Yu 'noyle' Liu, All Rights Reserved
 *
 */

// Make sure we can't access this file directly from the browser.
if(!defined('IN_MYBB'))
{
	die('This file cannot be accessed directly.');
}

define('CLOUDFLARE_HELPER_PLUGIN_REQUIRE', dirname(__FILE__).'/cloudflare_helper.php');
define('CLOUDFLARE_ENHANCER_PLUGIN', __FILE__);

if(!defined('CLOUDFLARE_HELPER_PLUGIN'))
{
	if(cloudflare_enhancer_check_dependency())
	{
		require_once CLOUDFLARE_HELPER_PLUGIN_REQUIRE;
	}
}

/**
 * Inline styles for showing a debug info below MyBB's <debugstuff>.
 */
define('CLOUDFLARE_ENHANCER_DEBUG_INLINE_STYLE', "<style>/*#debug:after {clear: right;}*/ #cloudflare_enhancer_debug {clear: right; float: right; text-align: right; font-size: 11px;}</style>");

if(cloudflare_enhancer_check_dependency())
{
	cloudflare_enhancer_add_hook_to_debugstuff();
}

function cloudflare_enhancer_info()
{
	$plugin_info = array(
		'name'			=> 'Cloudflare Enhancer',
		'description'	=> 'A very primitive plugin for exploiting HTTP headers set by Cloudflare. Plugin <strong>Cloudflare Helper</strong> is required!',
		'website'		=> 'https://github.com/yuliu/mybb-plugin-cloudflare-helper',
		'author'		=> 'Yu \'noyle\' Liu',
		'authorsite'	=> 'https://github.com/yuliu/mybb-plugin-cloudflare-helper',
		'version'		=> '0.1',
		'compatibility'	=> '18*',
		'codename'		=> 'noyle-cloudflare_enhancer'
	);

	if(!cloudflare_enhancer_check_dependency())
	{
		$plugin_info['description'] .= "<br /><span style=\"font-weight: bold;\">Error: plugin <a href=\"https://github.com/yuliu/mybb-plugin-cloudflare-helper\" target=\"_blank\"><em>Cloudflare Helper</em></a> is not detected. <a href=\"https://github.com/yuliu/mybb-plugin-cloudflare-helper/issues\" target=\"_blank\">Click here for help!</a></span>";
	}
	return $plugin_info;
}

function cloudflare_enhancer_check_dependency()
{
	if(!defined('CLOUDFLARE_HELPER_PLUGIN'))
	{
		if(file_exists(CLOUDFLARE_HELPER_PLUGIN_REQUIRE))
		{
			require_once CLOUDFLARE_HELPER_PLUGIN_REQUIRE;
			return true;
		}
		else
		{
			return false;
		}
	}
	return true;
}

function cloudflare_enhancer_add_hook_to_debugstuff()
{
	global $mybb_plugin_cloudflare_helper;
	$cf_cc = $mybb_plugin_cloudflare_helper->get_country_code();

	global $plugins;

	if(!empty($cf_cc) && function_exists('cloudflare_enhancer_hook_'.$cf_cc))
	{
		$plugins->add_hook('cloudflare_helper_global_start_'.$cf_cc, 'cloudflare_enhancer_hook_'.$cf_cc);
	}
	else if($mybb_plugin_cloudflare_helper->is_from_tor())
	{
		$plugins->add_hook('cloudflare_helper_global_start_end_FROMTOR', 'cloudflare_enhancer_hook_FROMTOR');
	}
	else
	{
		$plugins->add_hook('cloudflare_helper_global_start_NOTFROMCF', 'cloudflare_enhancer_hook_NOTFROMCF');
	}

	$plugins->add_hook('pre_output_page', 'cloudflare_enhancer_hook_debug_style');
}

function cloudflare_enhancer_hook_debug_style(&$content)
{
	global $mybb;
	if($mybb->settings['extraadmininfo'] != 0 && ($mybb->usergroup['cancp'] == 1 || $mybb->dev_mode == 1))
	{
		$inline_styles = CLOUDFLARE_ENHANCER_DEBUG_INLINE_STYLE."\n</head>";
		$content = str_replace('</head>', $inline_styles, $content);
	}
}

function cloudflare_enhancer_hook_US()
{
	global $mybb, $plugins;
	if($mybb->settings['extraadmininfo'] != 0 && ($mybb->usergroup['cancp'] == 1 || $mybb->dev_mode == 1))
	{
		$plugins->add_hook('pre_output_page', 'cloudflare_enhancer_print_country_welcome_US');
	}
}

function cloudflare_enhancer_hook_CN()
{
	global $mybb, $plugins;
	if($mybb->settings['extraadmininfo'] != 0 && ($mybb->usergroup['cancp'] == 1 || $mybb->dev_mode == 1))
	{
		$plugins->add_hook('pre_output_page', 'cloudflare_enhancer_print_country_welcome_CN');
	}
}

function cloudflare_enhancer_hook_NOTFROMCF()
{
	global $mybb, $plugins;
	if($mybb->settings['extraadmininfo'] != 0 && ($mybb->usergroup['cancp'] == 1 || $mybb->dev_mode == 1))
	{
		$plugins->add_hook('pre_output_page', 'cloudflare_enhancer_print_country_welcome_NOTFROMCF');
	}
}

function cloudflare_enhancer_hook_FROMTOR()
{
	global $mybb, $plugins;
	if($mybb->settings['extraadmininfo'] != 0 && ($mybb->usergroup['cancp'] == 1 || $mybb->dev_mode == 1))
	{
		$plugins->add_hook('pre_output_page', 'cloudflare_enhancer_print_country_welcome_FROMTOR');
	}
}

function cloudflare_enhancer_print_country_welcome_US(&$content)
{
	$welcome = "<debugstuff>\n<div id=\"cloudflare_enhancer_debug\">Welcome, explorer from the United States!</div>";
	$content = str_replace('<debugstuff>', $welcome, $content);
}

function cloudflare_enhancer_print_country_welcome_CN(&$content)
{
	$welcome = "<debugstuff>\n<div id=\"cloudflare_enhancer_debug\">Welcome, my friend from the east!</div>";
	$content = str_replace('<debugstuff>', $welcome, $content);
}

function cloudflare_enhancer_print_country_welcome_NOTFROMCF(&$content)
{
	$welcome = "<debugstuff>\n<div id=\"cloudflare_enhancer_debug\">The site is currently not served through Cloudflare.<br />You're accessing it directly from its origin server!</div>";
	$content = str_replace('<debugstuff>', $welcome, $content);
}

function cloudflare_enhancer_print_country_welcome_FROMTOR(&$content)
{
	$welcome = "<debugstuff>\n<div id=\"cloudflare_enhancer_debug\">Welcome, stranger from the void!<br />You're accessing it from a Tor network.</div>";
	$content = str_replace('<debugstuff>', $welcome, $content);
}


Upload/inc/plugins/cloudflare_helper/cloudflare_helper.php
<?php
/**
 * MyBB 1.8 plugin: Clouldflare Helper
 * Website: https://github.com/yuliu/mybb-plugin-cloudflare-helper
 * License: https://github.com/yuliu/mybb-plugin-cloudflare-helper/blob/master/LICENSE
 * Copyright Yu 'noyle' Liu, All Rights Reserved
 *
 */

// Make sure we can't access this file directly from the browser.
if(!defined('IN_MYBB'))
{
	die('This file cannot be accessed directly.');
}

/**
 * Cloudflare IP ranges.
 * See: https://www.cloudflare.com/ips/
 */
define('CLOUDFLARE_HELPER_CF_IPV4ADDR_URL', 'https://www.cloudflare.com/ips-v4');
define('CLOUDFLARE_HELPER_CF_IPV6ADDR_URL', 'https://www.cloudflare.com/ips-v6');
define('CLOUDFLARE_HELPER_CF_IPV4ADDR_FILE', dirname(__FILE__).'/ips-v4.txt');
define('CLOUDFLARE_HELPER_CF_IPV6ADDR_FILE', dirname(__FILE__).'/ips-v6.txt');

/**
 * Country code in ISO 3166-1 Alpha 2 format
 * See: https://support.cloudflare.com/hc/en-us/articles/200168236-What-does-Cloudflare-IP-Geolocation-do-
 * See: https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
 */

/**
 * Cloudflare-defined: XX is used for clients without country code data.
 */
define('CLOUDFLARE_HELPER_CC_NODATA', 'XX');

/**
 * Cloudflare-defined: T1 is used for clients using the Tor network.
 */
define('CLOUDFLARE_HELPER_CC_TOR', 'T1');


class CloudflareHelper
{
	private $ready = false;
	private $via_cf_network = false;
	private $through_cf_network = false;
	
	private $cf_added_headers = array(
		'cf_ipcountry' => false,
		'cf_connecting_ip' => false,
		'x_forwarded_for' => false,
		'x_forwarded_proto' => false,
		'cf_ray' => false,
		'cf_visitor' => false,
		'cdn_loop' => false,
	);
	
	private $from_tor = false;
	private $cc = false;
	
	private $ips_v4 = '';
	private $ips_v6 = '';
	
	public function __construct()
	{
		$this->_load_cf_ips();
		$this->_init();
	}
	
	public function get_ip_country()
	{
		$this->ready ? 0 : $this->_init();
		return $this->is_via_cf_network() ? $this->cf_added_headers['cf_ipcountry'] : false;
	}
	
	public function get_country_code()
	{
		$this->ready ? 0 : $this->_init();
		return $this->is_via_cf_network() ? $this->cc : false;
	}
	
	public function get_country()
	{
		$this->ready ? 0 : $this->_init();
		//
	}
	
	public function get_connnecting_ip()
	{
		$this->ready ? 0 : $this->_init();
		return $this->is_via_cf_network() ? $this->cf_added_headers['cf_connecting_ip'] : false;
	}
	
	public function is_via_cf_network()
	{
		$this->ready ? 0 : $this->_init();
		return $this->via_cf_network;
	}
	
	public function is_through_cf_network()
	{
		$this->ready ? 0 : $this->_init();
		return $this->through_cf_network;
	}
	
	/**
	 * Get if a request is made from a Tor network via Cloudflare network.
	 * @return int|boolean 1 if the client is using Tor and the request is via Cloudflare, 0 if not using Tor but via Cloudflare. false if request is not via Cloudfalre network;
	 */
	public function is_from_tor()
	{
		$this->ready ? 0 : $this->_init();
		return $this->is_via_cf_network() ? $this->from_tor : false;
	}
	
	/**
	 * Get HTTP headers added by Cloudflare.
	 * @return array|boolean An array containing relevant HTTP headers; false if the request is not via Cloudflare network.
	 */
	public function get_cf_headers()
	{
		$this->ready ? 0 : $this->_init();
		if(!$this->is_via_cf_network())
		{
			return false;
		}
		return $this->cf_added_headers;
	}
	
	/**
	 * Check if current request is made via Cloudflare by checking if the IP in HTTP header 'remote_addr' is in the Cloudflare network.
	 * @return boolean true if the request is made via Cloudflare, false otherwise.
	 */
	private function _check_remote_address_via_cf_network()
	{
		if(empty($this->ips_v4) || empty($this->ips_v6))
		{
			return false;
		}
		
		$remote_ip = $_SERVER["REMOTE_ADDR"];
		foreach(array_merge($this->ips_v4, $this->ips_v6) as $cf_ips_cidr)
		{
			$cf_ip_range = fetch_ip_range($cf_ips_cidr);
			$packed_remote_ip = my_inet_pton($remote_ip);
			
			if(is_array($cf_ip_range))
			{
				if(strcmp($cf_ip_range[0], $packed_remote_ip) <= 0 && strcmp($cf_ip_range[1], $packed_remote_ip) >= 0)
				{
					return true;
				}
			}
			elseif($remote_ip == $cf_ips_cidr)
			{
				return true;
			}
		}
		return false;
	}
	
	private function _init()
	{
		if(!$this->_check_remote_address_via_cf_network())
		{
			$this->via_cf_network = false;
			$this->ready = true;
			return;
		}

		$this->via_cf_network = true;
		$this->through_cf_network = true;
		$this->ready = true;

		$this->cf_added_headers['cf_ipcountry'] = isset($_SERVER["HTTP_CF_IPCOUNTRY"]) ? $_SERVER["HTTP_CF_IPCOUNTRY"] : '';
		$this->cf_added_headers['cf_connecting_ip'] = isset($_SERVER["HTTP_CF_CONNECTING_IP"]) ? $_SERVER["HTTP_CF_CONNECTING_IP"] : '';
		$this->cf_added_headers['cf_ray'] = isset($_SERVER["HTTP_CF_RAY"]) ? $_SERVER["HTTP_CF_RAY"] : '';
		$this->cf_added_headers['cf_visitor'] = isset($_SERVER["HTTP_CF_VISITOR"]) ? $_SERVER["HTTP_CF_VISITOR"] : '';
		$this->cf_added_headers['cdn_loop'] = isset($_SERVER["HTTP_CDN_LOOP"]) ? $_SERVER["HTTP_CDN_LOOP"] : '';

		$this->cf_added_headers['x_forwarded_for'] = isset($_SERVER["HTTP_X_FORWARDED_FOR"]) ? $_SERVER["HTTP_X_FORWARDED_FOR"] : '';
		$this->cf_added_headers['x_forwarded_proto'] = isset($_SERVER["HTTP_X_FORWARDED_PROTO"]) ? $_SERVER["HTTP_X_FORWARDED_PROTO"] : '';
		
		$this->from_tor = $this->cf_added_headers['cf_ipcountry'] == CLOUDFLARE_HELPER_CC_TOR ? 1 : 0;
		$this->cc = $this->cf_added_headers['cf_ipcountry'] == CLOUDFLARE_HELPER_CC_NODATA || $this->cf_added_headers['cf_ipcountry'] == CLOUDFLARE_HELPER_CC_TOR ? '' : $this->cf_added_headers['cf_ipcountry'];
	}
	
	private function _load_cf_ips_from_file($file)
	{
		if(!file_exists($file))
		{
			return false;
		}
		$ips = file_get_contents($file);
		if($ips !== false)
		{
			$ips = str_replace("\r", "\n", $ips);
			$ips = array_unique(array_filter(explode("\n", $ips)));
			return $ips;
		}
		else
		{
			return false;
		}
	}
	
	private function _load_cf_ips()
	{
		$ips = $this->_load_cf_ips_from_file(CLOUDFLARE_HELPER_CF_IPV4ADDR_FILE);
		if($ips !== false)
		{
			$this->ips_v4 = $ips;
		}
		
		$ips = $this->_load_cf_ips_from_file(CLOUDFLARE_HELPER_CF_IPV6ADDR_FILE);
		if($ips !== false)
		{
			$this->ips_v6 = $ips;
		}
		
		// TODO: read/store IPs from/into MyBB's data_cache.
		$cached_data = array(
			'timestamp' => TIME_NOW,
			'ips_v4' => $this->ips_v4,
			'ips_v6' => $this->ips_v6,
		);
	}
	
	public function cloudflare_helper_add_hooks_global_start()
	{
		$this->ready ? 0 : $this->_init();
		
		global $plugins;
		$plugins->run_hooks('cloudflare_helper_global_start');
		
		if(!$this->via_cf_network)
		{
			$plugins->run_hooks('cloudflare_helper_global_start_NOTFROMCF');
		}
		else if($this->from_tor === 1)
		{
			$plugins->run_hooks('cloudflare_helper_global_start_FROMTOR');
		}
		else if(empty($this->cc))
		{
			$plugins->run_hooks('cloudflare_helper_global_start_NOCOUNTRY');
		}
		else
		{
			$plugins->run_hooks('cloudflare_helper_global_start_'.$this->cc);
		}
	}
	
	public function cloudflare_helper_add_hooks_global_start_end()
	{
		$this->ready ? 0 : $this->_init();
		
		global $plugins;
		$plugins->run_hooks('cloudflare_helper_global_start_end');
		
		if(!$this->via_cf_network)
		{
			$plugins->run_hooks('cloudflare_helper_global_start_end_NOTFROMCF');
		}
		else if($this->from_tor === 1)
		{
			$plugins->run_hooks('cloudflare_helper_global_start_end_FROMTOR');
		}
		else if(empty($this->cc))
		{
			$plugins->run_hooks('cloudflare_helper_global_start_end_NOCOUNTRY');
		}
		else
		{
			$plugins->run_hooks('cloudflare_helper_global_start_end_'.$this->cc);
		}
	}
}

